stardist/bioimageio_utils.py

```diff
@@ -0,0 +1,266 @@
+    axes_norm = set(axes_net).intersection(set(axes_check_and_normalize(test_input_norm_axes, disallowed='S')))
+    axes_norm = ''.join(a for a in axes_net if a in axes_norm) # preserve order of axes_net
+    axes_norm_num = tuple(axes_net.index(a) for a in axes_norm)
+
+    # normalize input image
+    test_input_norm = normalize(test_input, pmin=min_percentile, pmax=max_percentile, axis=axes_norm_num)
+
+    net_axes_in = axes_net.lower()
+    net_axes_out = axes_check_and_normalize(model._axes_out).lower()
+    ndim_tensor = len(net_axes_out) + 1
+
+    input_min_shape = list(axes_net_div_by)
+    input_min_shape[axes_net.index('C')] = model.config.n_channel_in
+    input_step = list(axes_net_div_by)
+    input_step[axes_net.index('C')] = 0
+
+    if mode == "keras_hdf5":
+        output_names = ("prob", "dist") + (("class_prob",) if model._is_multiclass() else ())
+        output_n_channels = (1, model.config.n_rays,) + ((1,) if model._is_multiclass() else ())
+        output_scale = [1]+list(1/g for g in model.config.grid) + [0]
+
+    elif mode == "tensorflow_saved_model_bundle":
+        if model._is_multiclass():
+            raise NotImplementedError("Tensorflow SavedModel not supported for multiclass models yet")
+        # regarding input/output names: https://github.com/CSBDeep/CSBDeep/blob/b0d2f5f344ebe65a9b4c3007f4567fe74268c813/csbdeep/utils/tf.py#L193-L194
+        input_names  = ["input"]
+        output_names = ["output"]
+        output_n_channels = (1 + model.config.n_rays,)
+        output_scale = [1]*(ndim_tensor-1) + [0]
+
+    # TODO need config format that is compatible with deepimagej; discuss with Esti
+    # TODO do we need parameters for down/upsampling here?
+    metadata, _ = _import()
+    package_data = metadata("stardist")
+    config = dict(
+        stardist=dict(
+            stardist_version=package_data["Version"],
+            thresholds=dict(nms=model.thresholds.nms, prob=model.thresholds.prob)
+        )
+    )
+
+    n_inputs = len(input_names)
+    assert n_inputs == 1
+    input_axes = "b" + net_axes_in.lower()
+    input_config = dict(
+        input_name=input_names,
+        input_min_shape=[[1]+input_min_shape],
+        input_step=[[0]+input_step],
+        input_axes=[input_axes],
+        input_data_range=[["-inf", "inf"]],
+        preprocessing=[dict(scale_range=dict(
+            mode="per_sample",
+            axes=axes_norm.lower(),
+            min_percentile=min_percentile,
+            max_percentile=max_percentile,
+        ))]
+    )
+
+    n_outputs = len(output_names)
+    output_axes = "b" + net_axes_out.lower()
+    output_config = dict(
+        output_name=output_names,
+        output_data_range=[["-inf", "inf"]] * n_outputs,
+        output_axes=[output_axes] * n_outputs,
+        output_reference=[input_names[0]] * n_outputs,
+        output_scale=[output_scale] * n_outputs,
+        output_offset=[[0]*(ndim_tensor)] * n_outputs,
+    )
```

Comment:
Does bioimage.io supported tiled processing? If so, which parameters are relevant for that, especially to specify the overlap between neighboring tiles?

Reply:
Yes, the relevant parameter is `halo`.

Reply:
Thanks, I've found in the [documentation](https://github.com/bioimage-io/spec-bioimage-io/blob/gh-pages/model_spec_latest.md):
> The halo to crop from the output tensor (for example to crop away boundary effects or for tiling).

Does that mean the final output tensor will also be cropped, i.e. the output image will be smaller than the input image?

Reply:
no. halo is merely a hint for the caller of the model

the description of `halo` also says: 
> The halo is not cropped by the bioimage.io model, but is left to be cropped by the consumer software.
> Use `shape:offset` if the model output itself is cropped and input and output shapes not fixed.

We should make that clearer.

Reply:
That's even more confusing to me.

What I think is important: The prediction output should be the same, whether tiling is used or not (StarDist and CSBDeep work that way).

Reply:
> What I think is important: The prediction output should be the same, whether tiling is used or not (StarDist and CSBDeep work that way).

Yes, that's how it is done in bioimageio.core as well.

Reply:
I have added a halo to the model now in ef3e48f. For now, I just set 32 pixel halo (= overlap). Feel free to change this if you have a better heuristic @uschmidt83.